Device Firmware Update Cookbook (#109)
memfault/interrupt: _drafts/firmware-update-architecture.md

{% blockdiag size:120x40 %}
blockdiag {
    span_width = 100;
    // Set labels to nodes.
    C [label = "Bootloader"];
    A [label = "App Loader"];
    B [label = "Application"];
    C -> A [label = "Loads", fontsize=8];
    A -> B [label = "Ld, Updt", fontsize=8];

    E [label = "Updater"];
    A -> E [label = "Ld, Updt", fontsize=8];
    E -> A [label = "Updates", fontsize=8];
    C -> E [label = "Loads", style=dashed, fontsize=8];

    group {
        label = "Slot 0";
        color = "PaleGreen";
        C;
    }
    group {
        label = "Slot 1";
        color = "LightPink";
        A;
    }
    group {
        label = "Slot 2";
        color = "LemonChiffon";
        B; E;
    }
}
{% endblockdiag %}{:.diag4}

## Implementation Cookbook

### Image Metadata

Firmware images typically need to be bundled alongside some metadata. It
identifies what type of image it is, how large it is, what version it is, ...
etc.

That metadata must be:
1. Easy to find and parse for an external program
2. Accessible from within the firmware itself

My favorite approach is to add a struct at the beginning of the image to serve
as a header. Here's what it looks like:

```c
// image.h

> typedef struct __attribute__((packed)) {

Reviewer: build_id !!!! (and link your post in a comment)

  2019-08-13-how-to-write-a-bootloader-from-scratch %})
* [How to Write Linker Scripts for Firmware]({% post_url
  2019-06-25-how-to-write-linker-scripts-for-firmware %})
* [GNU Build IDs for Firmwre]({% post_url 2019-05-29-gnu-build-id-for-firmware
  %})
* [Building a Tiny CLI Shell for Tiny Firmware]({% post_url
  2020-06-09-firmware-shell %})

### Image Metadata

Firmware images typically need to be bundled alongside some metadata. It
identifies what type of image it is, how large it is, what version it is, ...
etc.

That metadata must be:
1. Easy to find and parse for an external program
2. Accessible from within the firmware itself

My favorite approach is to add a struct at the beginning of the image to serve
as a header. Here's what it looks like:

```c
// image.h
typedef struct __attribute__((packed)) {
    uint16_t image_magic;
    uint16_t image_hdr_version;
    uint32_t crc;
    uint32_t data_size;
    uint8_t image_type;
    uint8_t version_major;
    uint8_t version_minor;
    uint8_t version_patch;
    uint32_t vector_addr;
    uint32_t reserved;
    char git_sha[8];
} image_hdr_t;
```

The contents of that header will vary per application, but you must include:

> 1. A way to indicate that the header is valid (here we used a constant in

Reviewer 2: this didn't seem to format correctly

> `image_magic`)

Reviewer 2: Maybe we should explicitly call out what a "magic" value is / it's purpose. I know it's kinda ubiquitous but if you aren't use to looking at binary formats perhaps a new concept.

2. A version for the header, and
3. The address of the start of your vector table.

Each of our images must fill in its own header. For example, here it is for my
Application:

```c
// app.c
image_hdr_t image_hdr __attribute__((section(".image_hdr"))) = {
    .image_magic = IMAGE_MAGIC,
    .image_hdr_version = IMAGE_VERSION_CURRENT,
    .image_type = IMAGE_TYPE_APP,

>     .version_major = 1,

Reviewer 2: maybe worth explicitly adding the other fields and mentioning they get populated as part of the build process

```
// populated as part of a post compilation step
.crc = 0,
.data_size = 0,
```

```
// app.ld
SECTIONS
{
    .image_hdr : {
        KEEP (*(.image_hdr))
    } > slot2rom

    /* Vector table must be 128-bit aligned */
    . = ALIGN(128);

    .text : {
        *(.vectors)    /* Vector table */
        *(.text*)    /* Program code */
        . = ALIGN(4);
        *(.rodata*)    /* Read-only data */
        . = ALIGN(4);
    } >slot2rom

    ...
};
```

Some data can simply be hardcoded into the header. In my case, this includes the
`image_magic`, the semantic version, the image type, the header version, and the
vector table address. Other bits must be read from the environment and injected
as `DEFINES` by our Makefile. For example, this is how we get the `GIT_SHA`:

```makefile
# Makefile
GIT_SHA := \"$(shell $(GIT) rev-parse --short HEAD)\"
DEFINES += GIT_SHA=$(GIT_SHA)
CFLAGS += $(foreach d,$(DEFINES),-D$(d))
```

The `GIT_SHA` macro can then be called in our code to get the short (7-char)
hash as a string.

Some header information can only be calculated *after* the firmware has been
compiled and linked. For example, the length of the binary, its CRC, or its
cryptographic signature. To add it to the header, we must edit the binary
directly.

> Typically, I would write a Python script to do this but in a pinch you can use

Reviewer 2: should be able to replace with the python script at https://github.com/memfault/interrupt/commit/fc1065bc555b6fdb6f9c3340841b6368b4287e7d

some shell commands in a Makefile. For example, given `app.elf` I can do the
following to get a `bin` file with the CRC added:

```shell
# Create "app.data", a binary object without our image header
$ arm-none-eabi-objcopy app.elf app.data -O binary --remove-section=.image_hdr
# Compute the CRC and write it to a binary file
$ crc32 app.data | xxd -r -p | xxd -p -c1 | tac | xxd -r -p > app.crc
# Create "app.bin", the full firmware binary
$ arm-none-eabi-objcopy app.elf app.bin -O binary
# Inject the CRC into "app.bin"
$ dd if=app.crc of=app.bin obs=1 seek=4 conv=notrunc
```

That's it! We've added the CRC to our `bin` file, the main complexity really is
the Endian swap. We can now display the information in our firmware:

```c
// app.c
printf("App STARTED - version %d.%d.%d (%s) - CRC 0x%lx\n",
       image_hdr.version_major,
       image_hdr.version_minor,
       image_hdr.version_patch,
       image_hdr.git_sha,
       image_hdr.crc);
```

And as expected we get:

```
App STARTED - version 1.0.1 (e8e8d53) - CRC 0xeb4f7f52
```

### Loading Images

Both the Bootloader and the Loader need to load firmware images and start them.
This includes:

> 1. Locating them

Reviewer 2: this didn't format as a list either, I'm not sure what jekyll wants for that

    printf("Booting slot %d\n", slot);
    image_start(hdr);
}

// No image found, go into DFU mode
```

Here our approach to locating images is simple: we use hardcoded "slots" where
we expect images. Slot 1 is at `0x8004000`, and slot 2 is at `0x8020000`. We
simply iterate through the slots looking for valid headers.

Remember that the first field of our image header is the "magic" value. In
`image_get_header` we read the header and verify the magic value.

```c
// image.c
const image_hdr_t *image_get_header(image_slot_t slot) {
    const image_hdr_t *hdr = NULL;

    switch (slot) {
        case IMAGE_SLOT_1:
            hdr = (const image_hdr_t *)&__slot1rom_start__;
            break;
        case IMAGE_SLOT_2:
            hdr = (const image_hdr_t *)&__slot2rom_start__;
            break;
        default:
            break;
    }

    if (hdr && hdr->image_magic == IMAGE_MAGIC) {
        return hdr;
    } else {
        return NULL;
    }
}
```

To validate the image, we compute its CRC and compare it with the value written
in the header. This will allow you to catch data corruption or bit errors, and
save you some trouble down the line. Note that verifying CRC on boot is not

> appropriate for every applicaton, as it can slow down boot significantly.

Reviewer 2: could also call out its not appropriate for security reasons either / allude to a future post on the how to do signing

the vector table address has moved), and branching to it with a bit of assembly.

```c
static void prv_start_image(void *pc, void *sp) {
    __asm("           \n\
          msr msp, r1 \n\
          bx r0       \n\
    ");
}

void image_start(const image_hdr_t *hdr) {
    const vector_table_t *vectors = (const vector_table_t *)hdr->vector_addr;
    SCB_VTOR = (uint32_t)vectors;
    prv_start_image(vectors->reset, vectors->initial_sp_value);
     __builtin_unreachable();
}
```

If no image is found, the loader should go into DFU mode and waits for a new image to
be transferred.

> Note: The `VTOR` register is available on the Cortex-M0+,M3,M4,M7, but not on
> the older M0. Unfortunately this makes building a bootloader much more
> complicated on that platform.

### Writing & Committing Images

Writing a new firmware image is the whole purpose of this exercise, so how do we
do the deed? We won't cover *transferring* the image over to your firmware,
that's varies greatly depending on your use, but once the transfer has occured
we must:

1. Write the image
2. Verify it, and
3. Mark it as valid.

At the high level, here's our flow:

```c

> uint8_t *data_ptr = FIRMWARE_DATA;

Reviewer 2: maybe replace FIRMWARE_DATA define with function signature

```
int perform_update(void *fw_payload, size_t fw_payload_len) {
   uint8_t *data_ptr = fw_payload;
...
}
```

image_hdr_t *hdr = (image_hdr_t *)data_ptr;
data_ptr += sizeof(image_hdr_t);

if (dfu_write_data(IMAGE_SLOT_2,
                   data_ptr,
                   build_fwup_example_app_bin_len)) {
    return -1;
}

if (image_validate(IMAGE_SLOT_2, hdr)) {
    return -1;
};

if (dfu_commit_image(IMAGE_SLOT_2, hdr)) {
    return -1;
};

scb_reset_system();
```

> Writing the data is relatively straightforward. For flash memory You'll

Reviewer 2: ```suggestion
Writing the data is relatively straightforward. For flash memory you'll
```

typically erase the releavant sector, then use the `page_program` function
provided with your flash driver to write new data. Make sure however that you do
not write the image header during this process. The header will only be written
during the "commit" phase.

> We've covered verifying in image in the previous section. Here again we may

Reviewer 2: ```suggestion
We've covered how to verify an image in the previous section. Here again we may
```

check the CRC, or verify a cryptographic signature.

If at any point the update is interrupted, the header will be missing and our
Loader will refuse to load image. Once everything checks out, we *commit* the
image by writing the header. This marks the image as valid and allows it to be
loaded.

Here is my implementation of the `commit` step, it's nothing too complicated:

```c
int dfu_commit_image(image_slot_t slot, const image_hdr_t *hdr) {
    uint32_t addr = (uint32_t)(slot == IMAGE_SLOT_1 ?
                               &__slot1rom_start__ : &__slot2rom_start__);

    uint8_t *data_ptr = (uint8_t *)hdr;
    for (int i = 0; i < sizeof(image_hdr_t); ++i) {
        flash_program_byte(addr + i, data_ptr[i]);
    }

    return 0;
}
```

> Just as writing the header is marks an image as valid, we can invalidate an

Reviewer 2: ```suggestion
Just as writing the header marks an image as valid, we can invalidate an
```

image simply by zeroing it out. This is much faster than erasing the whole
flash, and won't leave us in a half-working state (e.g. having erased only half
of the image).

```c
int dfu_invalidate_image(image_slot_t slot) {
    // We just write 0s over the image header
    uint32_t addr = (uint32_t)(slot == IMAGE_SLOT_1 ?
                               &__slot1rom_start__ : &__slot2rom_start__);
    for (int i = 0; i < sizeof(image_hdr_t); ++i) {
        flash_program_byte(addr + i, 0);
    }

    return 0;
}
```

### Shared Memory

It is often useful to exchange information between our different programs. The
application may want to signal to the loader that it should go into DFU mode on
the next reboot, and the loader may want to pass arguments to the application.

On some platforms, scratch registers can be used as mailboxes between the
different programs. This is the case for the STM32. For all others we must make
do with RAM. "But François, isn't RAM volatile"? Not quite: as long as your MCU
remains powered, the SRAM will hold its state. For short durations, you can

> treat is as non-volatile storage.

Reviewer 2: ```suggestion
treat it as non-volatile storage.
```

The simplest way to carve out an area of RAM to be shared across our programs is

> to declare it in a shared linker script. We've covered linker scripts on

Reviewer 2: could link to article here

Interrupt, so I won't detail the syntax again. This is what it looks like:

```
// memory_map.ld
MEMORY
{
  sharedram (rwx): ORIGIN = 0x20000000, LENGTH = 256
  ram (rwx)      : ORIGIN = 0x20000100, LENGTH = 128K - LENGTH(sharedram)
}

SECTIONS
{
    .shared_memory (NOLOAD) : {
        KEEP(*(.shared_memory))
    } >sharedram
}
```

This shared linker script can then be included by the individual program's
linker script with the `INCLUDE` directive.

Now that we have a region of RAM defined, we can define a structure for it. You
will want that structure to be packed to guard against alignment mismatches
across your programs as you update compiler flags and versions. Here again we
use the `section` attribut to assign our symbol to the `.shared_memory` section.

```c
// shared_memory.c
typedef struct __attribute__((packed)) {
    uint32_t magic;
    uint32_t flags;
} shared_memory_t;

shared_memory_t shared_memory __attribute__((section(".shared_memory")));
```

> In the event of power loss, RAM will lost its state. To detect those events we

Reviewer 2: ```suggestion
In the event of power loss, RAM will lose its state. To detect those events we
```

use a magic value here as well. On boot, we check the magic number and
initialize the struct if it isn't what we'd expect.

```c
void shared_memory_init(void) {

>     if (shared_memory.magic != MAGIC) {

Reviewer 2: Some missed opportunities for spelling things in ascii :p 

`const uint32_t MAGIC = 0xbadcafe;` ->

`const uint32_t MAGIC = 0x4d524853;`

```
xxd &shared_memory 4
00000000: 5348 524d                                SHRM
```

        printf("Shared memory uinitialized, setting magic\n");
        memset(&shared_memory, 0, sizeof(shared_memory_t));
        shared_memory.magic = MAGIC;
    }
}
```

A common use case for shared memory is to signal to the Loader that we should
not load the Application on the next boot and instead go into DFU mode. We use a
simple flag to signal this:

```c
bool shared_memory_is_dfu_requested(void) {
    return (shared_memory.flags & DFU_REQUESTED) != 0;
}

void shared_memory_set_dfu_requested(bool yes) {
    shared_memory.flags |= DFU_REQUESTED;
}
```

> Using the shell from Tyler's last post, I created a shell command in the

Reviewer 2: could link here

app if the counter exceeds a threshold:

```c
const int max_boot_attempts = 3;
if (shared_memory_get_boot_counter() >= max_boot_attempts) {
    shared_memory_clear_boot_counter();
    printf("App unstable, dropping back into DFU mode\n");
    break;
}
```

Otherwise, it increments the counter and starts the application.

```c
printf("Booting slot 2\n");
shared_memory_increment_boot_counter();
image_start(hdr);
```

Last but not least, the Application must clear the counter once it believes it
is stable. When might that be? That depends on your application. One option is
to set a timer and mark the app as stable after it has been running for a few
minutes. Another option is to wait until it has managed to connect to a gateway
or the cloud. Whatever you do, the main requirement is that the Application can
receive an instruction to boot into DFU mode, so that we can bail out if it
turns out there is another bug lurking.

In my example, I simply created a firmware shell command which marks the app as
stable:

```c
int cli_command_mark_stable(int argc, char *argv[]) {
    shell_put_line("Marking app as stable");
    shared_memory_clear_boot_counter();
    return 0;
}
```

> 

Reviewer 2: ```suggestion
{:.no_toc}
```
